### PR TITLE
[AOTAutograd / Functionalization] Fix incorrect `expand_copy_inverse`

### LINKED ISSUE #114302: Mutation after `tensor.expand` returns wrong result.
### 🐛 Describe the bug

```python
import torch

@torch.compile(backend="inductor")
def foo(x):
  y = x.expand(2, *x.shape)
  y[0, 0] = 5
  return y

device = "cpu"
x = torch.arange(6, device=device)
r = foo(x)

print(x)
print(r)
```

```python
# Expected (Eager CPU)
tensor([ 5,  1, 2, 3, 4, 5])
tensor([[ 5,  1, 2, 3, 4, 5],
        [ 5, 1, 2, 3, 4, 5]])

# Actual
tensor([ 5,  2,  4,  6,  8, 10])
tensor([[ 5,  2,  4,  6,  8, 10],
        [ 5,  2,  4,  6,  8, 10]])
```

### Versions

PyTorch version: 2.2.0a0+git2bf3ca1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Fedora Linux 38 (Workstation Edition) (x86_64)
GCC version: (conda-forge gcc 13.2.0-2) 13.2.0
Clang version: Could not collect
CMake version: version 3.27.6
Libc version: glibc-2.37

Python version: 3.10.0 | packaged by conda-forge | (default, Nov 20 2021, 02:24:10) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-6.5.10-200.fc38.x86_64-x86_64-with-glibc2.37
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 9 5900X 12-Core Processor
CPU family:                         25
Model:                              33
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU(s) scaling MHz:                 71%
CPU max MHz:                        4950.1948
CPU min MHz:                        2200.0000
BogoMIPS:                           7399.77
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           6 MiB (12 instances)
L3 cache:                           64 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.1.0
[pip3] dalle2-pytorch==1.14.2
[pip3] ema-pytorch==0.3.0
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy==1.6.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] onnx==1.15.0
[pip3] open-clip-torch==2.23.0
[pip3] optree==0.9.1
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.3.5
[pip3] torch==2.2.0a0+gited447b4
[pip3] torch-fidelity==0.3.0
[pip3] torch_geometric==2.4.0
[pip3] torch-xla==2.2.0
[pip3] torchaudio==2.2.0a0+d9942ba
[pip3] torchbench==0.1
[pip3] torchdata==0.7.0a0+ce07539
[pip3] torchmetrics==1.2.0
[pip3] torchmultimodal==0.1.0b0
[pip3] torchtext==0.16.0a0+45e4b8c
[pip3] torchvision==0.17.0a0+af3077e
[pip3] vector-quantize-pytorch==1.11.3

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/114302

Only partial fix, will be correct in all eager passing cases, but will fail to throw error when eager throws error.

---

Deferring to @bdhirsh on follow up on fixing the case when eager throws an error. IMO it's not a priority and I'm OK with not fixing.

---

If in long term we want to fix, I think it can be handled with an analysis similar to https://github.com/pytorch/pytorch/blob/bbdd9b059fa06404701f0d39abaae75ebab5d27f/torch/_functorch/aot_autograd.py#L2202

We can check if expand dims are simultaneously mutated after analyzing in an fx pass.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_aot_autograd.py [hunk 1]
```diff
@@ -992,6 +993,36 @@ def f(x):
                 ),
             )
 
+    def test_aot_autograd_expand_mutation_functionalizes(self):
```
Comment (jon-chuang) [inline test/dynamo/test_aot_autograd.py:996]: could you add another test that the backwards is correct?

**Comment by jon-chuang** [comment]
This PR:

```python
INFO: TRACED GRAPH
 ===== Forward graph 0 =====
 <eval_with_key>.2 from /home/jonch/Desktop/Programming/mlsys/pytorch/torch/fx/experimental/proxy_tensor.py:509 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f16[6]"):
        # File: /home/jonch/Desktop/sdpa.py:2197, code: y = x.expand(3, *x.shape) #expand(1, *x.shape)
        expand: "f16[3, 6]" = torch.ops.aten.expand.default(arg0_1, [3, 6])
        
        # File: /home/jonch/Desktop/sdpa.py:2198, code: y[0:3, 0].add_(5)
        select: "f16[3]" = torch.ops.aten.select.int(expand, 1, 0);  expand = None
        add: "f16[3]" = torch.ops.aten.add.Tensor(select, 5);  select = None
        expand_1: "f16[3, 6]" = torch.ops.aten.expand.default(arg0_1, [3, 6])
        select_scatter: "f16[3, 6]" = torch.ops.aten.select_scatter.default(expand_1, add, 1, 0);  expand_1 = add = None
        sub: "f16[3, 6]" = torch.ops.aten.sub.Tensor(select_scatter, arg0_1);  select_scatter = None
        sum_1: "f16[1, 6]" = torch.ops.aten.sum.dim_IntList(sub, [0], True);  sub = None
        view: "f16[6]" = torch.ops.aten.view.default(sum_1, [6]);  sum_1 = None
        add_1: "f16[6]" = torch.ops.aten.add.Tensor(arg0_1, view);  arg0_1 = view = None
        
        # No stacktrace found for following nodes
        expand_3: "f16[3, 6]" = torch.ops.aten.expand.default(add_1, [3, 6])
        return (add_1, expand_3)
```

Main:

```python
INFO: TRACED GRAPH
 ===== Forward graph 0 =====
 <eval_with_key>.4 from /home/jonch/Desktop/Programming/mlsys/pytorch/torch/fx/experimental/proxy_tensor.py:509 in wrapped class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "i64[6]"):
        # File: /home/jonch/Desktop/sdpa.py:2197, code: y = x.expand(3, *x.shape)
        expand: "i64[3, 6]" = torch.ops.aten.expand.default(arg0_1, [3, 6])
        
        # File: /home/jonch/Desktop/sdpa.py:2198, code: y[0, 0] = 0
        _tensor_constant0 = self._tensor_constant0
        lift_fresh_copy: "i64[]" = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0);  _tensor_constant0 = None
        select: "i64[6]" = torch.ops.aten.select.int(expand, 0, 0);  expand = None
        select_1: "i64[]" = torch.ops.aten.select.int(select, 0, 0);  select = None
        copy: "i64[]" = torch.ops.aten.copy.default(select_1, lift_fresh_copy);  select_1 = lift_fresh_copy = None
        expand_1: "i64[3, 6]" = torch.ops.aten.expand.default(arg0_1, [3, 6])
        select_2: "i64[6]" = torch.ops.aten.select.int(expand_1, 0, 0)
        select_scatter: "i64[6]" = torch.ops.aten.select_scatter.default(select_2, copy, 0, 0);  select_2 = copy = None
        select_scatter_1: "i64[3, 6]" = torch.ops.aten.select_scatter.default(expand_1, select_scatter, 0, 0);  expand_1 = select_scatter = None
        sum_1: "i64[1, 6]" = torch.ops.aten.sum.dim_IntList(select_scatter_1, [0], True);  select_scatter_1 = None
        view: "i64[6]" = torch.ops.aten.view.default(sum_1, [6]);  sum_1 = None
        
        # No stacktrace found for following nodes
        copy_: "i64[6]" = torch.ops.aten.copy_.default(arg0_1, view);  arg0_1 = None
        expand_3: "i64[3, 6]" = torch.ops.aten.expand.default(view, [3, 6]);  view = None
        return (expand_3,)
```

**Review by bdhirsh — APPROVED** [review]

**Comment by bdhirsh** [comment]
@jon-chuang looks like my late review is causing you merge conflicts. Let me know if/when you have a chance to fix. Otherwise, I can make a new PR with your changes
